File: deep_learning_go/internal/nn/affine.go
package nn

import (
	"deep_learning_go/internal/mat/v2"
	"math"
)

type Affine struct {
	w  *mat.Mat
	b  *mat.Mat
	dW *mat.Mat
	dB *mat.Mat
	x  *mat.Mat
}

func XavierInit(shape ...int) *mat.Mat {
	scale := math.Sqrt(2.0 / float64(shape[0]+shape[1]))
	return mat.Random(shape).ScalarMul(scale)
}

func NewAffine(inputSize, outputSize int) *Affine {
	W := XavierInit(inputSize, outputSize)
	b := mat.MatZeros([]int{1, outputSize})

	return &Affine{
		w: W,
		b: b,
	}
}

func (a *Affine) Forward(x *mat.Mat) *mat.Mat {
	a.x = x

	return a.x.Mul(a.w).Add(a.b)
}

func (a *Affine) Backward(dout *mat.Mat) *mat.Mat {
	a.dW = a.x.T().Mul(dout)
	a.dB = dout.SumWithAxis(0, true)

	dx := dout.Mul(a.w.T())

	return dx
}

func (a *Affine) W() *mat.Mat {
	return a.w
}
func (a *Affine) B() *mat.Mat {
	return a.b
}
func (a *Affine) DW() *mat.Mat {
	return a.dW
}
func (a *Affine) DB() *mat.Mat {
	return a.dB
}

func (a *Affine) SetW(w *mat.Mat) {
	a.w = w
}
func (a *Affine) SetB(b *mat.Mat) {
	a.b = b
}
func (a *Affine) SetDW(dw *mat.Mat) {
	a.dW = dw
}
func (a *Affine) SetDB(db *mat.Mat) {
	a.dB = db
}

File: deep_learning_python/lib/affine.py
import numpy as np
class Affine:
    def __init__(self, input_size, output_size):
        self.W = np.random.randn(input_size, output_size) * np.sqrt(2.0 / input_size)
        self.b = np.zeros((1, output_size))
        self.dW = None
        self.db = None

    def forward(self, X):
        self.X = X
        return np.dot(X, self.W) + self.b

    def backward(self, d_out, train=True):
        if train:
            self.dW = np.dot(self.X.T, d_out)
            self.db = np.sum(d_out, axis=0, keepdims=True)

        dX = np.dot(d_out, self.W.T)
        return dX
